errors.go

```diff
@@ -36,3 +27,33 @@ var (
 	// due to floating-point arithmetic.
 	ErrNoProgress = errors.New("linesearch: no change in location after Linesearcher step")
 )
+
+// ErrFuncInit is returned when an initial function value is invalid. The error
+// state may be either +Inf or NaN. ErrFuncInit satisfies the error interface.
+type ErrFuncInit float64
+
+func (err ErrFuncInit) Error() string {
+	switch {
+	case math.IsInf(float64(err), 1):
+		return "optimize: initial function value is infinite"
+	case math.IsNaN(float64(err)):
+		return "optimize: initial function value is NaN"
+	default:
+		panic("optimize: bad ErrFuncInit")
+	}
+}
+
+// ErrGradInit is returned when an initial gradient is invalid. The error state
+// may be either ±Inf or NaN. ErrGradInit satisfies the error interface.
+type ErrGradInit float64
```

Comment:
Re. having Init in the name: I don't feel strongly either way. ErrFunc, ErrGrad are nice and short, but they suggest that they can be used for any invalid values returned from the objective function  at any point while (at least at the moment) they are specific to initial data. For example Problem.Status might want to return them. Of course, in principle ErrFunc etc can do it but I think it should not be necessary. All this optimization stuff we have at the moment assumes well behaved functions, so that if we start from valid initial data, there always is a sufficiently small neighborhood where the function also returns valid data. If not, we can backtrack closer to previous location and retry until we eventually make no progress and then ErrNoProgress. We do not do it yet, but we should. Fixing Backtracking to do it is trivial, Bisection, mmmm, not sure off the top of my head. Anyway, I'm deviating from the topic.

Re. one ErrInit type: Nice idea. One type, one Error method. Perhaps I am slightly leaning towards this side? We could abuse Operation to hold where the invalid value comes from but it seems that the fields should not be exported in this case.

Also, now we are checking only the value and the gradient. Should we check also the Hessian? Consistency says so, but what if one day we have Newton for "large" problems? Also, do you plan to check also the constraints?